### DISCUSSION TITLE
Row manipulation

### DISCUSSION THREAD

**Discussion by PradipKumarDas** [biolab/orange3#5497] [Q&A]
Hello.

Thanks for building and maintaining this beautiful piece of software called **Orange**!

I am new to it and I was trying to learn while building workflow while building actual solution on Anaconda platform for one of the 
Data Science projects. I thought it this way because I believed it is good way to not just learn it but also to know if Orange can solve not just academic projects, but also professional projects.

While completing the data workflow through few of the basic widgets as part of Exploratory Data Analysis (EDA) in Orange, my requirement was to remove all rows from Data Table against a condition, and I struggled there to find a way, and finally landed on  Orange 3 documentation page https://orange3.readthedocs.io/projects/orange-data-mining-library/en/latest/reference/data.table.html that indicates that Row Manipulation methods were deprecated from version 3.24.

I understand that there might be reasons for that behind this decision, but I would expect users might be having their own reasons to have such requirements to play with row manipulations on data.

Hence, I would wish if this functionalities come back soon.

In the meantime, could you please if there are any alternative approaches to remove data from Data Table? Widget "Select Rows" is not of help here.

I have been using Orange version 3.26.

Regards,
Pradip Kumar Das

**Comment by ajdapretnar** [comment]
Could I ask why is Select Rows not ok here? What is its limitation?
The deprecation you are describing, from what I know, just means one cannot change the table directly, which makes sense in Orange. You need to create a new table from existing one. Am I right, @markotoplak?

**Comment by janezd** [comment]
Row manipulation is not coming back; #5381 will make `Orange.data.Table` even immutable, unless explicitly "unlocked". Modifying tables that you don't own is inherently unsafe.

Consider tables like Python tuples or string. When you need to modify a tuple or string, you create a new one. When you need to modify a table, create a new table that contains the selected rows.

**Comment by PradipKumarDas** [comment]
> Could I ask why is Select Rows not ok here? What is its limitation?
> The deprecation you are describing, from what I know, just means one cannot change the table directly, which makes sense in Orange. You need to create a new table from existing one. Am I right, @markotoplak?

Thanks @ajdapretnar for responding. Yes, let me mention that brought me considering for row manipulation feature. The dataset I have been working on contain several columns and one such column is for income. It has appropriate values for almost 75% of the rows and have missing values for remaining rows. After realising about few of the income as outliers, rows having those outliers need to be removed. But other rows having missing values in that income column must stay intact in the dataset.

Upon not finding row manipulation functionality, I tried achieving this with widget "Select Rows", but found that it is filtering rows with missing values, too, when condition was something like "[income column] is at most [number]".

As you and @janezd have mentioned there could be some means to achieve this without row manipulation, but I may not know as I am new to Orange.

Could you please help suggesting an approach so that I can continue with Orange as far as my present project is concerned? Actually, I wish to continue my parallel journey with Orange as it is going to be very good tool for my academics, too, unless I face a roadblock.

Regards,
Pradip Kumar Das

**Comment by PradipKumarDas** [comment]
Hell @janezd 

> Row manipulation is not coming back; #5381 will make `Orange.data.Table` even immutable, unless explicitly "unlocked". Modifying tables that you don't own is inherently unsafe.
> 
> Consider tables like Python tuples or string. When you need to modify a tuple or string, you create a new one. When you need to modify a table, create a new table that contains the selected rows.

Hello @janezd.

I have just replied to @ajdapretnar 's response. Would you mind referring that?

Regards,
Pradip Kumar Das

**Comment by janezd** [comment]
See https://orange3.readthedocs.io/projects/orange-data-mining-library/en/latest/reference/data.table.html, probably `from_numpy`.

**Comment by PradipKumarDas** [comment]
> See https://orange3.readthedocs.io/projects/orange-data-mining-library/en/latest/reference/data.table.html, probably `from_numpy`.

I had already referred that and it is from where I came to know about the deprecation of the row manipulation functionalities. I referred it once again as you have mentioned, and I doubt I understood if anything from there can help resolving my query.

**Comment by markotoplak** [comment]
If you need to take a subset of rows, use indexing:

```
import Orange
iris = Orange.data.Table("iris")
iris_1_15_99 = iris[[1,15,99]]  # selects rows 1, 15, 99
```

If you wish to construct a new table (if you modify values), probably `from_numpy`, as janezd suggested, will be the easiest. For some introduction to handling data in Orange, see https://orange3.readthedocs.io/projects/orange-data-mining-library/en/latest/tutorial/data.html

I am moving this thread the discussion forum.
